File: engine/api/src/main/scala/pl/touk/nussknacker/engine/api/process/ExpressionConfig.scala
package pl.touk.nussknacker.engine.api.process

import pl.touk.nussknacker.engine.api.dict.DictDefinition
import pl.touk.nussknacker.engine.api.expression.ExpressionParser

import scala.util.matching.Regex

//TODO: how to make this config less spel-centric?, move globalImports and optimizeCompilation to spel configuration
case class ExpressionConfig(globalProcessVariables: Map[String, WithCategories[AnyRef]],
                            globalImports: List[WithCategories[String]],
                            additionalClasses: List[Class[_]] = List.empty,
                            languages: LanguageConfiguration = LanguageConfiguration.default,
                            optimizeCompilation: Boolean = true,
                            strictTypeChecking: Boolean = true,
                            // TODO After moving categories on root level of all objects, we should consider replacing
                            //      this map with list and adding dictId into DictDefinition. Then we will be sure that
                            //      DictInstance have the same dictId as DictDefinition
                            dictionaries: Map[String, WithCategories[DictDefinition]] = Map.empty,
                            hideMetaVariable: Boolean = false,
                            strictMethodsChecking: Boolean = true,
                            staticMethodInvocationsChecking: Boolean = true,
                            methodExecutionForUnknownAllowed: Boolean = false,
                            dynamicPropertyAccessAllowed: Boolean = false,
                            spelExpressionBlacklist: SpelExpressionBlacklist = SpelExpressionBlacklist.default
                           )

object ExpressionConfig {
  val empty = ExpressionConfig(Map.empty, Nil)
}

object LanguageConfiguration {
  val default = LanguageConfiguration(List.empty)
}

case class LanguageConfiguration(expressionParsers: List[ExpressionParser])


case class SpelExpressionBlacklist(blacklistedPatterns: Set[Regex])

object SpelExpressionBlacklist {

  val default: SpelExpressionBlacklist = SpelExpressionBlacklist(
    Set(
      "^(java.lang.System).*$".r,
      "^(java.net).*$".r,
      "^(java.io).*$".r,
      "^(java.nio).*$".r,
      "^(java.lang.invoke).*$".r,
      "^(java.lang.reflect).*$".r
    ))
}

File: engine/api/src/main/java/pl/touk/nussknacker/engine/javaapi/process/ExpressionConfig.java
package pl.touk.nussknacker.engine.javaapi.process;

import pl.touk.nussknacker.engine.api.dict.DictDefinition;
import pl.touk.nussknacker.engine.api.process.LanguageConfiguration;
import pl.touk.nussknacker.engine.api.process.WithCategories;
import scala.collection.immutable.List$;

import java.io.Serializable;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public class ExpressionConfig implements Serializable {

    private final Map<String, WithCategories<Object>> globalProcessVariables;

    private final List<WithCategories<String>> globalImports;

    private final List<Class<?>> additionalClasses;

    private final LanguageConfiguration languages;

    private final boolean optimizeCompilation;

    private final boolean strictTypeChecking;

    private final Map<String, WithCategories<DictDefinition>> dictionaries;

    private final boolean hideMetaVariable;

    private final boolean methodExecutionForUnknownAllowed;

    private final boolean dynamicPropertyAccessAllowed;

    public ExpressionConfig(Map<String, WithCategories<Object>> globalProcessVariables, List<WithCategories<String>> globalImports,  List<Class<?>> additionalClasses) {
        this(globalProcessVariables, globalImports, additionalClasses, new LanguageConfiguration(List$.MODULE$.empty()), true, true, Collections.emptyMap(), false, true, false);
    }

    public ExpressionConfig(Map<String, WithCategories<Object>> globalProcessVariables, List<WithCategories<String>> globalImports,
                            List<Class<?>> additionalClasses, LanguageConfiguration languages, boolean optimizeCompilation, boolean strictTypeChecking,
                            Map<String, WithCategories<DictDefinition>> dictionaries, boolean hideMetaVariable, boolean methodExecutionForUnknownAllowed,
                            boolean dynamicPropertyAccessAllowed) {
        this.globalProcessVariables = globalProcessVariables;
        this.globalImports = globalImports;
        this.additionalClasses = additionalClasses;
        this.languages = languages;
        this.optimizeCompilation = optimizeCompilation;
        this.strictTypeChecking = strictTypeChecking;
        this.dictionaries = dictionaries;
        this.hideMetaVariable = hideMetaVariable;
        this.methodExecutionForUnknownAllowed = methodExecutionForUnknownAllowed;
        this.dynamicPropertyAccessAllowed = dynamicPropertyAccessAllowed;
    }

    public Map<String, WithCategories<Object>> getGlobalProcessVariables() {
        return globalProcessVariables;
    }

    public List<WithCategories<String>> getGlobalImports() {
        return globalImports;
    }

    public List<Class<?>> getAdditionalClasses() {
        return additionalClasses;
    }

    public LanguageConfiguration getLanguages() {
        return languages;
    }

    public boolean isOptimizeCompilation() {
        return optimizeCompilation;
    }

    public boolean isStrictTypeChecking() {
        return strictTypeChecking;
    }

    public Map<String, WithCategories<DictDefinition>> getDictionaries() {
        return dictionaries;
    }

    public boolean isHideMetaVariable() {
        return hideMetaVariable;
    }

    public boolean isMethodExecutionForUnknownAllowed() {
        return methodExecutionForUnknownAllowed;
    }

    public boolean isDynamicPropertyAccessAllowed() {
        return dynamicPropertyAccessAllowed;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ExpressionConfig that = (ExpressionConfig) o;
        return optimizeCompilation == that.optimizeCompilation &&
                strictTypeChecking == that.strictTypeChecking &&
                Objects.equals(globalProcessVariables, that.globalProcessVariables) &&
                Objects.equals(globalImports, that.globalImports) &&
                Objects.equals(additionalClasses, that.additionalClasses) &&
                Objects.equals(languages, that.languages) &&
                Objects.equals(dictionaries, that.dictionaries) &&
                Objects.equals(hideMetaVariable, that.hideMetaVariable) &&
                Objects.equals(methodExecutionForUnknownAllowed, that.methodExecutionForUnknownAllowed) &&
                Objects.equals(dynamicPropertyAccessAllowed, that.dynamicPropertyAccessAllowed);
    }

    @Override
    public int hashCode() {
        return Objects.hash(globalProcessVariables, globalImports, additionalClasses, languages, optimizeCompilation,
                strictTypeChecking, dictionaries, hideMetaVariable, methodExecutionForUnknownAllowed, dynamicPropertyAccessAllowed);
    }

}
